mqtt/mqtt-edgehub/src/settings.rs

```diff
@@ -27,33 +26,65 @@ lazy_static! {
     };
 }
 
+/// `BrokerEnvironment` is our custom implementation of `config::Source`
+/// that can handle existing `EdgeHub` env settings and convert them
+/// into broker config structure.
 #[derive(Debug, Clone)]
 pub struct BrokerEnvironment;
 
-impl config::Source for BrokerEnvironment {
-    fn clone_into_box(&self) -> Box<dyn config::Source + Send + Sync> {
+impl Source for BrokerEnvironment {
+    fn clone_into_box(&self) -> Box<dyn Source + Send + Sync> {
         Box::new((*self).clone())
     }
 
-    // Currently, BrokerEnvironment allows only the following four environment variables to be set externally.
-    // Otherwise, all values must come from the default.json file
-    fn collect(&self) -> Result<HashMap<String, config::Value>, ConfigError> {
+    // Currently, BrokerEnvironment allows only the env variables explicitly
+    // defined in the method below.
+    //
+    // We use intermediate instance of `Config` to enumerate all env vars
+    // and then manually map them to our internal config structure.
+    // This is done for two reasons:
+    // - our broker config structure does not match legacy EdgeHub env vars,
+    // - `Config` does a bunch of useful things - takes care of
+    //   evn vars casing, prefixing, separators...
+    //
+    // NOTE: if adding new env vars - don't forget to use lowercase
+    // and update `check_env_var_name_override` test.
+    fn collect(&self) -> Result<HashMap<String, Value>, ConfigError> {
+        let mut host_env = Config::new();
+        // regular env vars
+        host_env.merge(Environment::new())?;
+        // broker specific vars
+        host_env.merge(Environment::with_prefix("MqttBroker_").separator(":"))?;
+        host_env.merge(Environment::with_prefix("MqttBroker_").separator("__"))?;
+
+        dbg!(host_env.collect()?);
+
         let mut result: HashMap<String, config::Value> = HashMap::new();
-        if let Ok(val) = env::var("mqttBroker__max_queued_messages") {
+
+        // session
+        if let Ok(val) = host_env.get::<String>("maxinflightmessages") {
```

Comment:
Can we get `Value` from the config instead of getting a string and then converting to the `Value`?

Reply:
good idea. It calls `T::deserialize` inside. I wonder how that would work if you call deserialize on self type :)